test/infrastructure/docker/docker/types/node.go

```diff
@@ -176,45 +175,82 @@ func (c *containerCmd) RunLoggingOutputOnFail(ctx context.Context) ([]string, er
 }
 
 func (c *containerCmd) Run(ctx context.Context) error {
-	args := []string{
-		"exec",
+	cli, err := dockerClient.NewClientWithOpts(dockerClient.FromEnv, dockerClient.WithAPIVersionNegotiation())
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	execConfig := &dockerTypes.ExecConfig{
 		// run with privileges so we can remount etc..
 		// this might not make sense in the most general sense, but it is
 		// important to many kind commands
-		"--privileged",
-	}
-	if c.stdin != nil {
-		args = append(args,
-			"-i", // interactive so we can supply input
-		)
+		Privileged:   true,
+		Cmd:          append([]string{c.command}, c.args...), // with the command specified
+		AttachStdout: true,
+		AttachStderr: true,
+		AttachStdin:  (c.stdin != nil), // interactive so we can supply input
 	}
+
 	// set env
-	for _, env := range c.env {
-		args = append(args, "-e", env)
-	}
-	// specify the container and command, after this everything will be
-	// args the the command in the container rather than to docker
-	args = append(
-		args,
-		c.nameOrID, // ... against the container
-		c.command,  // with the command specified
-	)
-	args = append(
-		args,
-		// finally, with the caller args
-		c.args...,
-	)
-	cmd := exec.CommandContext(ctx, "docker", args...)
-	if c.stdin != nil {
-		cmd.SetStdin(c.stdin)
-	}
-	if c.stderr != nil {
-		cmd.SetStderr(c.stderr)
-	}
-	if c.stdout != nil {
-		cmd.SetStdout(c.stdout)
-	}
-	return errors.WithStack(cmd.Run())
+	execConfig.Env = append(execConfig.Env, c.env...)
+
+	response, err := cli.ContainerExecCreate(ctx, c.nameOrID, *execConfig)
+	if err != nil {
+		return errors.WithStack(err)
+	}
+
+	execID := response.ID
+	if execID == "" {
+		return errors.New("exec ID empty")
+	}
+
+	resp, err := cli.ContainerExecAttach(ctx, execID, dockerTypes.ExecStartCheck{})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer resp.Close()
+
+	// If there is input, send it through to its stdin
+	inputDone := make(chan struct{})
+	go func() {
+		if c.stdin != nil {
+			_, _ = io.Copy(resp.Conn, c.stdin)
+			_ = resp.CloseWrite()
+		}
+		close(inputDone)
+	}()
+
+	// Read out any output from the call
+	outputDone := make(chan error)
+	go func() {
+		if c.stdout != nil && c.stderr != nil {
```

Comment:
I think we can also move this if before the func.

I'm wondering if it's intended that we wait for the timeout when those fields are not set. Do we actually always need these fields set and should check this via a guard clause?